### PR TITLE
arch/ostest: remove CONFIG_TESTING_OSTEST_FPUSIZE & board level implement

### PR DESCRIPTION
**Pull request by anchao**
## Summary

board/arch_fpu*: move arch_[get|cmp]fpu to common arch
boards/ostest: remove board ostest implement
arch/arm: export arm_saveusercontext()

## Impact

N/A

## Testing

ostest

### CODE REVIEW

**Comment by anchao** [comment]
Depends on https://github.com/apache/incubator-nuttx-apps/pull/1141

File: include/nuttx/arch.h [hunk 1]
```diff
@@ -2563,6 +2563,16 @@ uint32_t up_perf_gettime(void);
 uint32_t up_perf_getfreq(void);
 void up_perf_convert(uint32_t elapsed, FAR struct timespec *ts);
 
+/****************************************************************************
+ * Name: up_getcontext
```
Comment (xiaoxiang781216) [inline include/nuttx/arch.h:2567]: up_getcontext->up_saveusercontext

File: include/nuttx/arch.h [hunk 2]
```diff
@@ -2563,6 +2563,16 @@ uint32_t up_perf_gettime(void);
...
+ *
+ * Description:
+ *   Save the current thread context
+ *
+ ****************************************************************************/
+
+int up_saveusercontext(uint32_t *saveregs);
```
Comment (xiaoxiang781216) [inline include/nuttx/arch.h:2574]: uint32_t *->FAR void *

File: include/nuttx/arch.h [hunk 3]
```diff
@@ -2573,6 +2573,20 @@ void up_perf_convert(uint32_t elapsed, FAR struct timespec *ts);
...
+/****************************************************************************
+ * Name: up_fpucmp
+ *
+ * Description:
+ *   compare FPU areas from thread context
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_FPU
+bool up_fpucmp(FAR const uint32_t *saveregs1, FAR const uint32_t *saveregs2);
```
Comment (xiaoxiang781216) [inline include/nuttx/arch.h:2585]: bool up_fpucmp(FAR const void *saveregs1, FAR const void *saveregs2);

File: arch/arm/include/armv7-m/irq_cmnvector.h [hunk 4]
```diff
@@ -133,6 +133,8 @@
 #define XCPTCONTEXT_REGS    (HW_XCPT_REGS + SW_XCPT_REGS)
 #define XCPTCONTEXT_SIZE    (4 * XCPTCONTEXT_REGS)
 
+#define FPU_CONTEXT_REGS    SW_FPU_REGS
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv7-m/irq_cmnvector.h:136]: let's directly use SW_FPU_REGS

File: arch/arm/include/armv8-m/irq_cmnvector.h [hunk 5]
```diff
@@ -139,6 +139,8 @@
 #define XCPTCONTEXT_REGS    (HW_XCPT_REGS + SW_XCPT_REGS)
 #define XCPTCONTEXT_SIZE    (4 * XCPTCONTEXT_REGS)
 
+#define FPU_CONTEXT_REGS    SW_FPU_REGS
```
Comment (xiaoxiang781216) [inline arch/arm/include/armv8-m/irq_cmnvector.h:142]: ditto

File: arch/arm/src/common/arm_fpucmp.c [hunk 6]
```diff
@@ -25,70 +25,32 @@
...
-
-  flags = enter_critical_section();
-  arm_saveusercontext(g_saveregs);
-
-  /* Return only the floating register values */
-
-  memcpy(fpusave, &g_saveregs[REG_S0], (4*SW_FPU_REGS));
-  leave_critical_section(flags);
+  return memcmp(&saveregs1[REG_S0],
+                &saveregs2[REG_S0], (4 * FPU_CONTEXT_REGS)) == 0;
```
Comment (xiaoxiang781216) [inline arch/arm/src/common/arm_fpucmp.c:54]: FPU_CONTEXT_REGS->SW_FPU_REGS

File: include/nuttx/arch.h (hunk 1 above)
Comment (anchao) [reply include/nuttx/arch.h]: Done

File: include/nuttx/arch.h (hunk 2 above)
Comment (anchao) [reply include/nuttx/arch.h]: Done

File: include/nuttx/arch.h (hunk 3 above)
Comment (anchao) [reply include/nuttx/arch.h]: Done

File: arch/risc-v/src/common/riscv_fpucmp.c [hunk 7]
```diff
@@ -37,58 +34,27 @@
...
+  uintptr_t regs1 = (uintptr_t)saveregs1;
+  uintptr_t regs2 = (uintptr_t)saveregs2;
 
-  /* Return only the floating register values */
-
-  memcpy(fpusave, &g_saveregs[INT_XCPT_REGS], (4*FPU_XCPT_REGS));
-  leave_critical_section(flags);
+  return memcmp((void *)(regs1 + INT_REG_SIZE * INT_XCPT_REGS),
+                (void *)(regs2 + INT_REG_SIZE * INT_XCPT_REGS),
+                INT_REG_SIZE * FPU_XCPT_REGS) == 0;
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_fpucmp.c:58]: INT_REG_SIZE * FPU_XCPT_REGS->FPU_REG_SIZE * FPU_XCPT_REGS

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@pkarashchenko do you have any concern about  this patch?

**Comment by pkarashchenko** [comment]
> @pkarashchenko do you have any concern about  this patch?

Sorry, I missed it. I will make a review tomorrow.

File: arch/arm/src/common/arm_fpucmp.c [hunk 8]
```diff
@@ -25,70 +25,35 @@
...
+bool up_fpucmp(FAR const void *saveregs1, FAR const void *saveregs2)
 {
-  irqstate_t flags;
-
-  /* Take a snapshot of the thread context right now */
-
-  flags = enter_critical_section();
-  arm_saveusercontext(g_saveregs);
+  FAR const uint32_t *regs1 = saveregs1;
+  FAR const uint32_t *regs2 = saveregs2;
```
Comment (pkarashchenko) [inline arch/arm/src/common/arm_fpucmp.c:54]: let's remove `FAR` as Mr. @xiaoxiang781216 is cleaning things up for ARM

File: arch/risc-v/src/common/riscv_fpucmp.c [hunk 9]
```diff
@@ -37,58 +34,27 @@
...
+ *   compare FPU areas from thread context
+ *
  ****************************************************************************/
 
-/* Given an array of size CONFIG_TESTING_OSTEST_FPUSIZE, this function will
- * return the current FPU registers.
- */
-
-void arch_getfpu(FAR uint32_t *fpusave)
+bool up_fpucmp(FAR const void *saveregs1, FAR const void *saveregs2)
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_fpucmp.c:51]: Let's remove `FAR` as it is not available for risc-v

Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_fpucmp.c:51]: Let's remove `FAR` as it is not available for risc-v

File: arch/risc-v/src/common/riscv_fpucmp.c [hunk 10]
```diff
@@ -37,58 +34,27 @@
...
-  flags = enter_critical_section();
-  riscv_saveusercontext(g_saveregs);
+  uintptr_t regs1 = (uintptr_t)saveregs1;
+  uintptr_t regs2 = (uintptr_t)saveregs2;
 
-  /* Return only the floating register values */
-
-  memcpy(fpusave, &g_saveregs[INT_XCPT_REGS], (4*FPU_XCPT_REGS));
-  leave_critical_section(flags);
+  return memcmp((void *)(regs1 + INT_REG_SIZE * INT_XCPT_REGS),
```
Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_fpucmp.c:56]: Maybe next can be considered as an option
```
  const uintptr_t *regs1 = saveregs1;
  const uintptr_t *regs2 = saveregs2;

  return memcmp(&regs1[INT_XCPT_REGS], &regs2[INT_XCPT_REGS],
                INT_REG_SIZE * FPU_XCPT_REGS) == 0;
```

Comment (pkarashchenko) [inline arch/risc-v/src/common/riscv_fpucmp.c:56]: Maybe next can be considered as an option
```
  uintptr_t *regs1 = (uintptr_t *)saveregs1;
  uintptr_t *regs2 = (uintptr_t *)saveregs2;

  return memcmp(regs1 + INT_XCPT_REGS),
       regs2 + INT_XCPT_REGS),
       sizeof(uintptr_t) * FPU_XCPT_REGS) == 0;
```

File: include/nuttx/arch.h [hunk 11]
```diff
@@ -2563,6 +2563,30 @@ uint32_t up_perf_gettime(void);
...
+ *
+ * Description:
+ *   compare FPU areas from thread context
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ARCH_FPU
+bool up_fpucmp(FAR const void *saveregs1, FAR const void *saveregs2);
+#else
+#define up_fpucmp(r1, r2) (1)
```
Comment (pkarashchenko) [inline include/nuttx/arch.h:2587]:
```suggestion
#define up_fpucmp(r1, r2) (true)
```

Comment (pkarashchenko) [inline include/nuttx/arch.h:2587]:
```suggestion
#define up_fpucmp(r1, r2) (true)
```

File: arch/arm/src/common/arm_fpucmp.c (hunk 8 above)
Comment (anchao) [reply arch/arm/src/common/arm_fpucmp.c]: Done

File: arch/risc-v/src/common/riscv_fpucmp.c (hunk 9 above)
Comment (anchao) [reply arch/risc-v/src/common/riscv_fpucmp.c]: Done

File: arch/risc-v/src/common/riscv_fpucmp.c (hunk 10 above)
Comment (anchao) [reply arch/risc-v/src/common/riscv_fpucmp.c]: Done

File: include/nuttx/arch.h (hunk 11 above)
Comment (anchao) [reply include/nuttx/arch.h]: Done

File: arch/risc-v/src/common/riscv_fpucmp.c (hunk 10 above)
Comment (anchao) [reply arch/risc-v/src/common/riscv_fpucmp.c]: Done

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
macOS CI always fail to:
" wget --quiet https://static.dev.sifive.com/dev-tools/riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-apple-darwin.tar.gz"
which is fixed by #6094, let's ignore this error temporarily to save the ci resource.

**Pull request merged**